### PR TITLE
Cypress  capacitive touch controls added

### PR DESCRIPTION
**Pull request by tymoteuszblochmobica**
#### Problem
Added Cypress capacitive touch control to CHIP applications

 #### Summary of Changes
 Capsense libraries
 Cypress drivers
 LockApp and  LigthingAPP

### CODE REVIEW

File: config/mbed/CMakeLists.txt [hunk 1]
```diff
@@ -273,3 +274,34 @@ target_include_directories(${APP_TARGET} PRIVATE
...
+   ${CAPSENSE_PATH}/cy_capsense_csx_v2.c
+   ${CAPSENSE_PATH}/cy_capsense_csd_v2.c
+   ${CAPSENSE_PATH}/cy_capsense_filter.c
+   ${CAPSENSE_PATH}/cy_capsense_generator_v3.c
+   ${CAPSENSE_PATH}/cy_capsense_processing.c
+   ${CAPSENSE_PATH}/cy_capsense_selftest.c
+   ${CAPSENSE_PATH}/cy_capsense_sensing_v2.c
+   ${CAPSENSE_PATH}/cy_capsense_sensing_v3.c
+   ${CAPSENSE_PATH}/cy_capsense_structure.c
+)
```
Comment (pan-) [inline config/mbed/CMakeLists.txt:303]: Can you create a library for it. We can put it in https://github.com/ARMmbed/mbed-os-cypress-capsense-button .

**Review by pan- — CHANGES_REQUESTED** [review]

File: .gitmodules [hunk 2]
```diff
@@ -86,3 +86,12 @@
...
 	branch = master
+[submodule "third_party/repo"]
+	path = third_party/repo
+	url = https://github.com/Infineon/capsense.git
+[submodule "third_party/third_party/capsense/repo"]
+	path = third_party/third_party/capsense/repo
+	url = https://github.com/Infineon/capsense.git
+[submodule "third_party/capsense/repo"]
+	path = third_party/capsense/repo
+	url = https://github.com/Infineon/capsense.git
```
Comment (ATmobica) [inline .gitmodules:97]: what happened here? Why are there so many new submodules?

File: examples/lock-app/mbed/main/AppTask.cpp [hunk 3]
```diff
@@ -57,7 +61,10 @@ static LEDWidget sLockLED(MBED_CONF_APP_LOCK_STATE_LED);
 
 static mbed::InterruptIn sLockButton(LOCK_BUTTON);
 static mbed::InterruptIn sFunctionButton(FUNCTION_BUTTON);
-
+#ifdef CAPSENSE_ENABLED
```
[the remark this answers is not in the document]
Comment (tymoteuszblochmobica) [reply examples/lock-app/mbed/main/AppTask.cpp:64]: Done

File: examples/platform/mbed/capsense/capsense.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    cybsp_init();
+
+    initialize_capsense_tuner();
+    sleep_manager_lock_deep_sleep();
+
+    cy_status status = Cy_CapSense_Init(&cy_capsense_context);
+
+    if (CY_RET_SUCCESS != status)
+    {
+        printf("CapSense initialization failed. Status code: %lu\r\n", (unsigned long) status);
```
[the remark this answers is not in the document]
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:164]: Done

File: examples/platform/mbed/capsense/capsense.h [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    void init();
+    static Capsense & getInstance()
+    {
+        static Capsense instance;
+        return instance;
+    }
+
+private:
+    Capsense() {}
+    ~Capsense() {}
```
[the remark this answers is not in the document]
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.h:50]: Done

File: .gitmodules (hunk 2 above)
Comment (tymoteuszblochmobica) [reply .gitmodules:97]: Mistake. To be removed

File: examples/platform/mbed/capsense/capsense.cpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
+
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:1]: Missing copyright information.

File: examples/platform/mbed/capsense/capsense.cpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+                                                (CY_SYSPM_SKIP_CHECK_FAIL | CY_SYSPM_SKIP_BEFORE_TRANSITION |
+                                                 CY_SYSPM_SKIP_AFTER_TRANSITION),
+                                                &callbackParams,
+                                                NULL,
+                                                NULL };
+
+cy_stc_scb_ezi2c_context_t ezi2c_context;
+cyhal_ezi2c_t sEzI2C;
+cyhal_ezi2c_slave_cfg_t sEzI2C_sub_cfg;
+cyhal_ezi2c_cfg_t sEzI2C_cfg;
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:39]: Do we need global variables here? Are they referenced anywhere outside this file/scope?
Consider adding 'static' to each of them or put them into unnamed namespace.

File: examples/platform/mbed/capsense/capsense.h [hunk 8]
```diff
@@ -0,0 +1,139 @@
+#include "cy_pdl.h"
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.h:1]: Missing copyright information.

File: examples/platform/mbed/capsense/capsense.cpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+void Capsense::RunCapSenseScan()
+{
+
+    Cy_CapSense_Wakeup(&cy_capsense_context);
+
+    if (CY_CAPSENSE_NOT_BUSY == Cy_CapSense_IsBusy(&cy_capsense_context))
+    {
+
+        Cy_CapSense_ScanAllWidgets(&cy_capsense_context);
+    }
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:58]: Remove some empty lines here.

File: examples/platform/mbed/capsense/capsense.cpp [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    if (result != CY_RSLT_SUCCESS)
+    {
+        printf("i2c error.\r\n");
+        handle_error();
+    }
+}
+
+void Capsense::ProcessTouchStatus()
+{
+
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:94]: Remove this empty line.

File: examples/platform/mbed/capsense/capsense.cpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+        {
+            prevSliderPos = currSliderPos;
+            Capsense::getInstance().on_slider(currSliderPos);
+        }
+    }
+}
+
+void Capsense::CapSense_InterruptHandler(void)
+{
+
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:144]: Remove this empty line.

File: examples/lighting-app/mbed/CMakeLists.txt [hunk 12]
```diff
@@ -34,7 +34,7 @@ target_include_directories(${APP_TARGET} PRIVATE
                            ${MBED_COMMON}/util/include
                            ${CHIP_ROOT}/src/app
 )
-
+
```
Comment (lukgni) [inline examples/lighting-app/mbed/CMakeLists.txt:37]: Remove extra space.

File: examples/lighting-app/mbed/CMakeLists.txt [hunk 13]
```diff
@@ -87,6 +87,7 @@ if("BlueNRG_MS" IN_LIST MBED_TARGET_LABELS)
             mbed-ble-blue_nrg
     )
 endif()
+
```
Comment (lukgni) [inline examples/lighting-app/mbed/CMakeLists.txt:90]: Remove this empty line.

File: examples/platform/mbed/capsense/capsense.cpp [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+using namespace std::literals::chrono_literals;
+
+#define SLIDER_NUM_TOUCH (1u)
+#define LED_OFF (1u)
+#define LED_ON (0u)
+
+#define CAPSENSE_SCAN_PERIOD_MS (20ms)
+#define EZI2C_INTERRUPT_PRIORITY (3u)
+
+#define TUNER_ENABLE (1u)
```
Comment (lukgni) [inline examples/platform/mbed/capsense/capsense.cpp:13]: Consider using `constexpr` instead of `#define`

**Review by lukgni — CHANGES_REQUESTED** [review]

File: examples/platform/mbed/capsense/capsense.cpp (hunk 6 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:1]: Done

File: examples/platform/mbed/capsense/capsense.cpp (hunk 7 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:39]: Done

File: examples/platform/mbed/capsense/capsense.h (hunk 8 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.h:1]: Done

File: examples/platform/mbed/capsense/capsense.cpp (hunk 9 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:58]: Done

File: examples/platform/mbed/capsense/capsense.cpp (hunk 10 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:94]: Done

File: examples/platform/mbed/capsense/capsense.cpp (hunk 11 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:144]: done

File: examples/lighting-app/mbed/main/AppTask.cpp [hunk 15]
```diff
@@ -192,7 +207,7 @@ void AppTask::LightingActionEventHandler(AppEvent * aEvent)
 {
     LightingManager::Action_t action = LightingManager::INVALID_ACTION;
     int32_t actor                    = 0;
-
+    uint8_t value;
```
[the remark this answers is not in the document]
Comment (tymoteuszblochmobica) [reply examples/lighting-app/mbed/main/AppTask.cpp:210]: done

File: examples/lighting-app/mbed/CMakeLists.txt (hunk 13 above)
Comment (tymoteuszblochmobica) [reply examples/lighting-app/mbed/CMakeLists.txt:90]: done

File: examples/platform/mbed/capsense/capsense.cpp (hunk 14 above)
Comment (tymoteuszblochmobica) [reply examples/platform/mbed/capsense/capsense.cpp:13]: This causing compiler error

File: .gitmodules [hunk 16]
```diff
@@ -86,3 +86,6 @@
 	path = third_party/wifi-ism43362/repo
 	url = https://github.com/ATmobica/wifi-ism43362.git
 	branch = master
+[submodule "third_party/mbed-os-cypress-capsense-button"]
+	path = third_party/mbed-os-cypress-capsense-button
```
Comment (pan-) [inline .gitmodules:90]: Last change, can you put the submodule into `third_party/mbed-os-cypress-capsense-button/repo` ? 
This also requires to rework https://github.com/ARMmbed/mbed-os-cypress-capsense-button so there is no `repo` folder and its content is at the root.

**Review by pan- — CHANGES_REQUESTED** [review]

File: config/mbed/CMakeLists.txt [hunk 17]
```diff
@@ -239,6 +239,7 @@ target_include_directories(chip INTERFACE
     ${CHIP_ROOT}/src/lib
     ${CHIP_ROOT}/src/lib/core
     ${CHIP_ROOT}/third_party/nlassert/repo/include
+    ${CHIP_ROOT}/third_party/capsense
```
Comment (pan-) [inline config/mbed/CMakeLists.txt:242]: I don't think it belong in the CMakeList.txt that build CHIP. It belongs in the CMakeLists.txt of each example. 
This also require update of the shell scripts that link sources in `third_party`.

**Review by pan- — CHANGES_REQUESTED** [review]

File: .gitmodules (hunk 16 above)
Comment (tymoteuszblochmobica) [reply .gitmodules]: Done

File: config/mbed/CMakeLists.txt (hunk 17 above)
Comment (tymoteuszblochmobica) [reply config/mbed/CMakeLists.txt]: Done

File: examples/lock-app/mbed/mbed_app.json [hunk 18]
```diff
@@ -41,11 +41,13 @@
...
                 "WHD_PRINT_DISABLE"
             ],
+            "target.components_add": [
+                "capsense"
+            ],
             "led-active-state": 1,
             "system-state-led": "LED1",
             "lock-state-led": "P9_6",
-            "lock-button": "USER_BUTTON",
-            "function-button": "P9_7"
```
Comment (lukgni) [inline examples/lock-app/mbed/mbed_app.json:48]: DITTO

File: examples/lighting-app/mbed/mbed_app.json [hunk 19]
```diff
@@ -40,11 +40,13 @@
...
                 "NL_ASSERT_EXPECT_FLAGS=NL_ASSERT_FLAG_LOG"
             ],
+            "target.components_add": [
+                "capsense"
+            ],
             "led-active-state": 0,
             "system-state-led": "LED1",
             "lighting-state-led": "P9_6",
-            "lighting-button": "USER_BUTTON",
-            "function-button": "P9_7"
```
Comment (lukgni) [inline examples/lighting-app/mbed/mbed_app.json:47]: If we remove "function-button" here, how does "MBED_CONF_APP_FUNCTION_BUTTON" definition is resolved in AppTask.cpp? Additionally, I don't like the idea where` "function-button"` is still available for disco target (even if we assigned "NC" to it) but we removed it from cypress board.

**Review by lukgni — CHANGES_REQUESTED** [review]

**Comment by lukgni** [comment]
PR rebased onto development branch.
Some small fixes added:
- fix issue with wrong active state for status led in lock-app example
- remove Thread dependency in lock-app/lighting-app examples
- use familiar naming scheme for capsense module in .gitmodule file.
- update capsense submodule revision to include bugfixing related to missing `TRACE_GROUP` definition.

I noticed that in lighting-app example we can only receive one ZCL message after being fully provisioned. This issue needs definitely more investigation (especially development branch checking as a first step)

**Comment by lukgni** [comment]
> PR rebased onto development branch.
> Some small fixes added:
> 
>     * fix issue with wrong active state for status led in lock-app example
> 
>     * remove Thread dependency in lock-app/lighting-app examples
> 
>     * use familiar naming scheme for capsense module in .gitmodule file.
> 
>     * update capsense submodule revision to include bugfixing related to missing `TRACE_GROUP` definition.
> 
> 
> I noticed that in lighting-app example we can only receive one ZCL message after being fully provisioned. This issue needs definitely more investigation (especially development branch checking as a first step)

Yep, the same behavior can be observed on `development` branch.

**Comment by pan-** [comment]
> Yep, the same behavior can be observed on development branch.

Hehe, we're in for a `git bisect`.

File: config/mbed/CMakeLists.txt [hunk 20]
```diff
@@ -419,4 +419,19 @@ if (CONFIG_CHIP_PW_RPC_LIGHTING_PROTO)
     )
 endif(CONFIG_CHIP_PW_RPC_LIGHTING_PROTO)
 
-endif(CONFIG_CHIP_PW_RPC)
\ No newline at end of file
+endif(CONFIG_CHIP_PW_RPC)
+if("capsense" IN_LIST MBED_TARGET_LABELS)
```
Comment (pan-) [inline config/mbed/CMakeLists.txt:423]: Not sure what was missed in my review by Tymoteusz but the capsense library build is part of the application, not the Matter library.

**Review by pan- — CHANGES_REQUESTED** [review]

**Review by pan- — APPROVED** [review]

**Pull request merged**